core/src/main/java/org/neo4j/graphalgo/core/heavyweight/HeavyCypherGraphFactory.java

```diff
@@ -0,0 +1,78 @@
+package org.neo4j.graphalgo.core.heavyweight;
+
+import org.neo4j.graphalgo.api.*;
+import org.neo4j.graphalgo.core.IdMap;
+import org.neo4j.graphalgo.core.NullWeightMap;
+import org.neo4j.graphalgo.core.WeightMap;
+import org.neo4j.graphalgo.core.utils.RawValues;
+import org.neo4j.kernel.internal.GraphDatabaseAPI;
+
+/**
+ * @author mknblch
+ */
+public class HeavyCypherGraphFactory implements GraphSource {
+
+    private static final int BATCH_SIZE = 100_000;
+    private static final int INITIAL_NODE_COUNT = 1_000_000;
+    private static final int DEGREE = 10;
+
+    private final GraphDatabaseAPI api;
+    private final GraphDefinition setup;
+
+    public HeavyCypherGraphFactory(
+            GraphDatabaseAPI api,
+            GraphDefinition setup) {
+        this.api = api;
+        this.setup = setup;
+    }
+
+    @Override
+    public Graph build() {
+        return build(BATCH_SIZE);
+    }
+
+    @SuppressWarnings("WeakerAccess")
+    Graph build(int batchSize) {
+        final IdMap idMap = new IdMap(INITIAL_NODE_COUNT);
+        boolean hasRelationshipWeights = setup.hasRelationshipWeight();
+
+        boolean hasNodeWeights = setup.hasNodeWeight();
+        final WeightMapping nodeWeights = hasNodeWeights ?
+                new WeightMap(INITIAL_NODE_COUNT, setup.nodeDefaultWeight()) :
+                new NullWeightMap(setup.nodeDefaultWeight());
+
+        boolean hasNodeValue = setup.hasNodeProperty();
+        final WeightMapping nodeProps = hasNodeValue ?
+                new WeightMap(INITIAL_NODE_COUNT, setup.nodeDefaultPropertyValue()) :
+                new NullWeightMap(setup.nodeDefaultPropertyValue());
+
+        // todo compiled
+        api.execute(setup.nodeStatement()).accept(row -> {
+            long id = row.getNumber("id").longValue();
+            idMap.add(id);
+            if (hasNodeWeights) nodeWeights.set(id, row.get("weight"));
+            if (hasNodeValue) nodeProps.set(id, row.get("value"));
+            return true;
+        });
+
+        int nodeCount = idMap.size();
+
+        final AdjacencyMatrix matrix = new AdjacencyMatrix(nodeCount);
+
+        final WeightMapping relWeigths = hasRelationshipWeights ?
+                new WeightMap(nodeCount * DEGREE, setup.relationDefaultWeight()) :
+                new NullWeightMap(setup.relationDefaultWeight());
+
+        api.execute(setup.relationshipStatement()).accept(row -> {
+            int source = idMap.mapOrGet(row.getNumber("source").longValue());
```

Comment:
Could we assume that the node statement always loads all nodes we need and use `get` here, skipping the rel or emit a warning, when `get` returns -1?